gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyCacheInitializationTest.java

```diff
@@ -0,0 +1,449 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization;
+
+import org.apache.tinkerpop.gremlin.process.traversal.Step;
+import org.apache.tinkerpop.gremlin.process.traversal.Traversal;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.DefaultGraphTraversal;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.__;
+import org.apache.tinkerpop.gremlin.process.traversal.step.LambdaHolder;
+import org.apache.tinkerpop.gremlin.process.traversal.step.TraversalParent;
+import org.apache.tinkerpop.gremlin.process.traversal.step.util.EmptyStep;
+import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalHelper;
+import org.apache.tinkerpop.gremlin.util.function.Lambda;
+import org.javatuples.Pair;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.__.where;
+
+/**
+ * Testing whether the cache for the FilterRankingStrategy is correctly initialized,
+ * which includes gathering/collecting the traversal parents and their children steps
+ * in the correct deep first manner.
+ */
+@RunWith(JUnit4.class)
+public class FilterRankingStrategyCacheInitializationTest {
+
+
+    @Test
+    public void test_collection_lambda_nested() {
+        // traversal
+        final Traversal<?, ?> traversal = __.V().as("n").
+                where(
+                        __.or(
+                                __.select("n").hasLabel("software"),
+                                __.select("n").map(Lambda.function("it.get().out()"))
+                        )
+                ).
+                select("n").by("name");
+
+        // expected order of traversal parents
+        final List<Traversal<?, ?>> collectedTraversalParents = Arrays.asList(
+                __.or(
+                        __.select("n").hasLabel("software"),
+                        __.select("n").map(Lambda.function("it.get().out()"))
+                ),
+                __.where(
+                        __.or(
+                                __.select("n").hasLabel("software"),
+                                __.select("n").map(Lambda.function("it.get().out()"))
+                        )
+                ));
+
+        // cache entries
+        final List<Pair<Boolean, Set<String>>> expectedTraversalParentCacheEntries = Arrays.asList(
+                new Pair<>(true, new HashSet<>(Collections.emptyList())),
```

Comment:
Why need to pass `Collections.emptyList()` to `HashSet` constructor?

Reply:
It is not necessary. I removed it in the new revision. Also changed some arguments to Optional to make it clear that they do not need to be provided (where we then just compare with the old cache initialization method).